### PR TITLE
ui-linked-field-renderers

### PR DESCRIPTION
**Pull request by andrewtibbetts**
added editable button labels, moved add row button below row list, minor associated spacing tweaks

### CODE REVIEW

File: packages/renderer-bootstrap4/renderers/table-linked-struct-field-renderer.tsx [hunk 1]
```diff
@@ -31,18 +35,18 @@ const Bootstrap4TableLinkedStructFieldRenderer = ({
           <td>{x || ' '}</td>
         ))}
         <td>
-          <div className={createCssClass(cssName, 'row', 'actions')}>
+          <div className="text-nowrap " + {createCssClass(cssName, 'row', 'actions')}>
```
Comment (gregbty) [inline packages/renderer-bootstrap4/renderers/table-linked-struct-field-renderer.tsx]: What the significance of this? Is this something that absolutely needs to be in the library's renderer or is this specific to the application that uses it.

Also for concatenating class names use `combineCssClasses`. There are examples of it being used in other places.

File: packages/renderer-bootstrap4/renderers/table-linked-struct-field-renderer.tsx [hunk 2]
```diff
@@ -31,18 +35,18 @@ const Bootstrap4TableLinkedStructFieldRenderer = ({
...
             <Bootstrap4ButtonRenderer
               classes="btn btn-link"
-              text="Edit"
+              text=editLabel
```
Comment (gregbty) [inline packages/renderer-bootstrap4/renderers/table-linked-struct-field-renderer.tsx]: This should be wrapped in brackets i.e. `{editLabel}`.

File: packages/renderer-bootstrap4/renderers/table-linked-struct-field-renderer.tsx [hunk 3]
```diff
@@ -31,18 +35,18 @@ const Bootstrap4TableLinkedStructFieldRenderer = ({
...
               onClick={() => onEdit(index)}
               disabled={readOnlyValues[index]}
             />
-            {removeButtonVisible && <span>|</span>}
+            {showActionSeparator && <span>|</span>}
             {removeButtonVisible && (
               <Bootstrap4ButtonRenderer
                 classes="btn btn-link"
-                text="Remove"
+                text=removeLabel
```
Comment (gregbty) [inline packages/renderer-bootstrap4/renderers/table-linked-struct-field-renderer.tsx]: Same as above.

File: packages/renderer-bootstrap4/renderers/table-linked-struct-field-renderer.tsx [hunk 4]
```diff
@@ -87,6 +81,16 @@ const Bootstrap4TableLinkedStructFieldRenderer = ({
           )}
         </tbody>
       </table>
+      <div className={createCssClass(cssName, 'controls')}>
+        {canAdd() && (
+          <Bootstrap4ButtonRenderer
+            classes="btn btn-link"
+            text=addLabel
```
Comment (gregbty) [inline packages/renderer-bootstrap4/renderers/table-linked-struct-field-renderer.tsx]: Same as above; should be wrapped in brackets.

File: packages/renderer-bootstrap4/renderers/table-linked-struct-field-renderer.tsx [hunk 5]
```diff
@@ -87,6 +81,16 @@ const Bootstrap4TableLinkedStructFieldRenderer = ({
           )}
         </tbody>
       </table>
+      <div className={createCssClass(cssName, 'controls')}>
```
Comment (gregbty) [inline packages/renderer-bootstrap4/renderers/table-linked-struct-field-renderer.tsx]: Why was this moved to underneath the table? If this is an application specific change, we can talk through how to create a custom renderer.

File: packages/renderer-bootstrap4/renderers/table-linked-struct-field-renderer.tsx [hunk 6]
```diff
@@ -19,6 +19,10 @@ const Bootstrap4TableLinkedStructFieldRenderer = ({
   onAdd,
   onEdit,
   onRemove,
+  addLabel,
```
Comment (gregbty) [inline packages/renderer-bootstrap4/renderers/table-linked-struct-field-renderer.tsx]: Where are these properties coming from?

**Comment by gregbty** [comment]
Most of the changes here, seem to be application specific changes (however that assumption may be incorrect). If that is the case, let's talk through how to create a custom renderer.

File: packages/renderer-bootstrap4/renderers/table-linked-struct-field-renderer.tsx (hunk 5 above)
Comment (andrewtibbetts) [reply packages/renderer-bootstrap4/renderers/table-linked-struct-field-renderer.tsx]: In general, I think it's a more intuitive place for the button was application-specific.

**Comment by andrewtibbetts** [comment]
Please hit me up on Slack to tell me more about this custom renderer of which you speak, cuz it sounds like it might make this whole PR unnecessary. :)

**Comment by gregbty** [comment]
I'm going to close this. Let's talk offline how to set up a custom renderer. In the future, I'll refine the documentation so that this is public knowledge.
